subscriber/common/job/controller.go

```diff
@@ -150,6 +166,51 @@ func NewController(params Params) *Controller {
 	return controller
 }
 
+func createEtcdServices(params Params) (services.Services, error) {
+	iopts := instrument.NewOptions().
+		SetZapLogger(params.ServiceConfig.Logger).
+		SetMetricsScope(params.ServiceConfig.Scope)
+
+	// etcd key format: prefix/${env}/namespace/service/instanceId
+	params.ServiceConfig.EtcdConfig.Service = fmt.Sprintf("%s/%s",
+		config.ActiveAresNameSpace, params.ServiceConfig.EtcdConfig.Service)
+	// create a config service client to access to the etcd cluster services.
+	csClient, err := params.ServiceConfig.EtcdConfig.NewClient(iopts)
+	if err != nil {
+		return nil, err
+	}
+
+	servicesClient, err := csClient.Services(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return servicesClient, nil
+}
+
+func registerHeartBeatService(params Params, servicesClient services.Services) error {
+	sid := services.NewServiceID().
+		SetEnvironment(params.ServiceConfig.EtcdConfig.Env).
+		SetZone(params.ServiceConfig.EtcdConfig.Zone).
+		SetName(params.ServiceConfig.EtcdConfig.Service)
+
+	pInstance := placement.NewInstance().SetID(params.ServiceConfig.Environment.InstanceID)
+
+	ad := services.NewAdvertisement().
+		SetServiceID(sid).
+		SetPlacementInstance(pInstance)
+
+	err := servicesClient.SetMetadata(sid, services.NewMetadata().
+		SetHeartbeatInterval(*params.ServiceConfig.HeartbeatConfig.Interval).
+		SetLivenessInterval(*params.ServiceConfig.HeartbeatConfig.Timeout))
+	if err != nil {
+		return err
+	}
+
+	err = servicesClient.Advertise(ad)
```

Comment:
can you explain a little on the usage of heartbeatservice here? to etcd? and what will do if heartbeat lost?

Reply:
heartbeatservice is an etcd lease node (kind of similar zk ephemeral node). The key will be removed if hb is lost.

etcd implements retry function internally (kind of curator functionality)